File: src/parser/ast.ts
/**
 * Abstract Syntax Tree (AST) node definitions for EzLang
 */

import { Position } from '../types';

/**
 * Base interface for all AST nodes
 */
export interface ASTNode {
  type: string;
  position?: Position;
}

/**
 * Root program node
 */
export interface Program extends ASTNode {
  type: 'Program';
  body: Statement[];
}

/**
 * Base interface for statements
 */
export interface Statement extends ASTNode {}

/**
 * Base interface for expressions
 */
export interface Expression extends ASTNode {}

/**
 * Variable declaration: var name = value
 */
export interface VariableDeclaration extends Statement {
  type: 'VariableDeclaration';
  name: string;
  initializer: Expression | null;
}

/**
 * Function declaration: function name(params) { body }
 */
export interface FunctionDeclaration extends Statement {
  type: 'FunctionDeclaration';
  name: string;
  parameters: string[];
  body: BlockStatement;
}

/**
 * If statement: if condition { consequent } else { alternate }
 */
export interface IfStatement extends Statement {
  type: 'IfStatement';
  condition: Expression;
  consequent: BlockStatement;
  alternate: BlockStatement | null;
}

/**
 * For loop: for variable in iterable { body }
 */
export interface ForStatement extends Statement {
  type: 'ForStatement';
  variable: string;
  iterable: Expression;
  body: BlockStatement;
}

/**
 * While loop: while condition { body }
 */
export interface WhileStatement extends Statement {
  type: 'WhileStatement';
  condition: Expression;
  body: BlockStatement;
}

/**
 * Return statement: return value
 */
export interface ReturnStatement extends Statement {
  type: 'ReturnStatement';
  value: Expression | null;
}

/**
 * Block statement: { statements }
 */
export interface BlockStatement extends Statement {
  type: 'BlockStatement';
  statements: Statement[];
}

/**
 * Expression statement
 */
export interface ExpressionStatement extends Statement {
  type: 'ExpressionStatement';
  expression: Expression;
}

/**
 * Listen statement: listen "event" (param) { body }
 */
export interface ListenStatement extends Statement {
  type: 'ListenStatement';
  event: string;
  parameter: string;
  body: BlockStatement;
}

/**
 * Use statement: use "module" as name
 */
export interface UseStatement extends Statement {
  type: 'UseStatement';
  module: string;
  alias: string;
}

/**
 * Import statement: import "path/to/file.ez"
 */
export interface ImportStatement extends Statement {
  type: 'ImportStatement';
  path: string;
}

/**
 * Natural Language: start bot with token "abc"
 */
export interface StartBotCommand extends Statement {
  type: 'StartBotCommand';
  token: Expression;
}

/**
 * Natural Language: when message says "abc"
 */
export interface WhenMessageCommand extends Statement {
  type: 'WhenMessageCommand';
  condition: Expression;
  body: BlockStatement;
}

/**
 * Natural Language: when command "ping" is used
 */
export interface WhenCommandUsedCommand extends Statement {
  type: 'WhenCommandUsedCommand';
  commandName: Expression;
  body: BlockStatement;
}

/**
 * Natural Language: load [npm|python] package "xyz" as xyz
 */
export interface LoadPackageCommand extends Statement {
  type: 'LoadPackageCommand';
  packageManager: 'npm' | 'python';
  packageName: string;
  alias: string;
}

/**
 * Natural Language: reply with "abc"
 */
export interface ReplyWithCommand extends Statement {
  type: 'ReplyWithCommand';
  message: Expression;
}

/**
 * Natural Language: reply with embed "Title" "Description"
 */
export interface ReplyWithEmbedCommand extends Statement {
  type: 'ReplyWithEmbedCommand';
  title: Expression;
  description: Expression;
}

/**
 * Discord send command: send channel message
 */
export interface SendCommand extends Statement {
  type: 'SendCommand';
  target: Expression;
  message: Expression;
}

/**
 * Discord reply command: reply message response
 */
export interface ReplyCommand extends Statement {
  type: 'ReplyCommand';
  target: Expression;
  message: Expression;
}

/**
 * Discord react command: react message emoji
 */
export interface ReactCommand extends Statement {
  type: 'ReactCommand';
  target: Expression;
  emoji: Expression;
}

/**
 * Natural Language: react with "emoji"
 */
export interface ReactWithCommand extends Statement {
  type: 'ReactWithCommand';
  emoji: Expression;
}

/**
 * Natural Language: send "message" to channel "id"
 */
export interface SendToCommand extends Statement {
  type: 'SendToCommand';
  message: Expression;
  target: Expression;
}

/**
 * Natural Language: set variable to expression
 */
export interface SetVariableCommand extends Statement {
  type: 'SetVariableCommand';
  name: string;
  value: Expression;
}

/**
 * Natural Language: when bot starts
 */
export interface WhenBotStartsCommand extends Statement {
  type: 'WhenBotStartsCommand';
  body: BlockStatement;
}

/**
 * Natural Language: when user joins server
 */
export interface WhenUserJoinsCommand extends Statement {
  type: 'WhenUserJoinsCommand';
  body: BlockStatement;
}

/**
 * Natural Language: when message starts with "text"
 */
export interface WhenMessageStartsWithCommand extends Statement {
  type: 'WhenMessageStartsWithCommand';
  condition: Expression;
  body: BlockStatement;
}

/**
 * Natural Language: when button "id" is clicked
 */
export interface WhenButtonClickedCommand extends Statement {
  type: 'WhenButtonClickedCommand';
  buttonId: Expression;
  body: BlockStatement;
}

/**
 * Moderation: ban user "name" [for "reason"]
 */
export interface BanCommand extends Statement {
  type: 'BanCommand';
  user: Expression;
  reason?: Expression;
}

/**
 * Moderation: kick user "name" [for "reason"]
 */
export interface KickCommand extends Statement {
  type: 'KickCommand';
  user: Expression;
  reason?: Expression;
}

/**
 * Moderation: timeout user "name" for <number> minutes
 */
export interface TimeoutCommand extends Statement {
  type: 'TimeoutCommand';
  user: Expression;
  duration: Expression;
}

/**
 * Roles: add role "Admin" to user "Bob"
 */
export interface AddRoleCommand extends Statement {
  type: 'AddRoleCommand';
  role: Expression;
  user: Expression;
}

/**
 * Roles: remove role "Muted" from user "Bob"
 */
export interface RemoveRoleCommand extends Statement {
  type: 'RemoveRoleCommand';
  role: Expression;
  user: Expression;
}

/**
 * Threads: create thread "Help"
 */
export interface CreateThreadCommand extends Statement {
  type: 'CreateThreadCommand';
  name: Expression;
}

/**
 * Voice: join voice channel "id"
 */
export interface JoinVoiceCommand extends Statement {
  type: 'JoinVoiceCommand';
  channel: Expression;
}

/**
 * Voice: play audio "music.mp3"
 */
export interface PlayAudioCommand extends Statement {
  type: 'PlayAudioCommand';
  file: Expression;
}

/**
 * Slash Commands: register slash command "name" with description "desc"
 */
export interface RegisterSlashCommand extends Statement {
  type: 'RegisterSlashCommand';
  name: Expression;
  description: Expression;
  options: SlashCommandOption[];
}

export interface SlashCommandOption {
  name: Expression;
  description: Expression;
  type: string; // "string", "number", "user", "role", "boolean"
  required: boolean;
}

/**
 * Components: reply with button "id" labeled "text" style "primary"
 */
export interface ReplyWithButtonCommand extends Statement {
  type: 'ReplyWithButtonCommand';
  message: Expression;
  buttonId: Expression;
  label: Expression;
  style: string;
}

/**
 * Components: reply with menu "id" with options [...]
 */
export interface ReplyWithMenuCommand extends Statement {
  type: 'ReplyWithMenuCommand';
  message: Expression;
  menuId: Expression;
  options: MenuOption[];
}

export interface MenuOption {
  label: Expression;
  value: Expression;
  description?: Expression;
}

/**
 * NLP Condition: if user has role "Admin"
 */
export interface IfUserHasRoleStatement extends Statement {
  type: 'IfUserHasRoleStatement';
  user: Expression;
  role: Expression;
  consequent: BlockStatement;
  alternate: BlockStatement | null;
}

/**
 * NLP Condition: if user has permission "BAN_MEMBERS"
 */
export interface IfUserHasPermissionStatement extends Statement {
  type: 'IfUserHasPermissionStatement';
  user: Expression;
  permission: Expression;
  consequent: BlockStatement;
  alternate: BlockStatement | null;
}

/**
 * Binary expression: left operator right
 */
export interface BinaryExpression extends Expression {
  type: 'BinaryExpression';
  operator: string;
  left: Expression;
  right: Expression;
}

/**
 * Unary expression: operator operand
 */
export interface UnaryExpression extends Expression {
  type: 'UnaryExpression';
  operator: string;
  operand: Expression;
}

/**
 * Call expression: callee(arguments)
 */
export interface CallExpression extends Expression {
  type: 'CallExpression';
  callee: Expression;
  arguments: Expression[];
}

/**
 * Member expression: object.property
 */
export interface MemberExpression extends Expression {
  type: 'MemberExpression';
  object: Expression;
  property: Expression;
  computed: boolean; // true for obj[prop], false for obj.prop
}

/**
 * Identifier
 */
export interface Identifier extends Expression {
  type: 'Identifier';
  name: string;
}

/**
 * Literal value
 */
export interface Literal extends Expression {
  type: 'Literal';
  value: string | number | boolean | null;
  raw: string;
}

/**
 * Array literal: [elements]
 */
export interface ArrayLiteral extends Expression {
  type: 'ArrayLiteral';
  elements: Expression[];
}

/**
 * Object literal: {key: value, ...}
 */
export interface ObjectLiteral extends Expression {
  type: 'ObjectLiteral';
  properties: ObjectProperty[];
}

/**
 * Object property in an object literal
 */
export interface ObjectProperty extends ASTNode {
  type: 'ObjectProperty';
  key: string;
  value: Expression;
}

/**
 * Assignment expression: target = value
 */
export interface AssignmentExpression extends Expression {
  type: 'AssignmentExpression';
  target: Identifier | MemberExpression;
  value: Expression;
}

/**
 * Type guard functions
 */
export function isExpression(node: ASTNode): node is Expression {
  return (
    node.type === 'BinaryExpression' ||
    node.type === 'UnaryExpression' ||
    node.type === 'CallExpression' ||
    node.type === 'MemberExpression' ||
    node.type === 'Identifier' ||
    node.type === 'Literal' ||
    node.type === 'ArrayLiteral' ||
    node.type === 'ObjectLiteral' ||
    node.type === 'AssignmentExpression'
  );
}

export function isStatement(node: ASTNode): node is Statement {
  return (
    node.type === 'VariableDeclaration' ||
    node.type === 'FunctionDeclaration' ||
    node.type === 'IfStatement' ||
    node.type === 'ForStatement' ||
    node.type === 'WhileStatement' ||
    node.type === 'ReturnStatement' ||
    node.type === 'BlockStatement' ||
    node.type === 'ExpressionStatement' ||
    node.type === 'ListenStatement' ||
    node.type === 'UseStatement' ||
    node.type === 'ImportStatement' ||
    node.type === 'SendCommand' ||
    node.type === 'ReplyCommand' ||
    node.type === 'ReactCommand' ||
    node.type === 'StartBotCommand' ||
    node.type === 'WhenMessageCommand' ||
    node.type === 'WhenCommandUsedCommand' ||
    node.type === 'LoadPackageCommand' ||
    node.type === 'ReplyWithCommand' ||
    node.type === 'ReplyWithEmbedCommand' ||
    node.type === 'ReactWithCommand' ||
    node.type === 'SendToCommand' ||
    node.type === 'SetVariableCommand' ||
    node.type === 'WhenBotStartsCommand' ||
    node.type === 'WhenUserJoinsCommand' ||
    node.type === 'WhenMessageStartsWithCommand' ||
    node.type === 'WhenButtonClickedCommand' ||
    node.type === 'BanCommand' ||
    node.type === 'KickCommand' ||
    node.type === 'TimeoutCommand' ||
    node.type === 'AddRoleCommand' ||
    node.type === 'RemoveRoleCommand' ||
    node.type === 'CreateThreadCommand' ||
    node.type === 'JoinVoiceCommand' ||
    node.type === 'PlayAudioCommand' ||
    node.type === 'RegisterSlashCommand' ||
    node.type === 'ReplyWithButtonCommand' ||
    node.type === 'ReplyWithMenuCommand' ||
    node.type === 'IfUserHasRoleStatement' ||
    node.type === 'IfUserHasPermissionStatement'
  );
}

File: Go-Standalone/internal/ast/ast.go
package ast

import "fmt"

type Node interface {
	TokenLiteral() string
}

type Statement interface {
	Node
	statementNode()
}

type Expression interface {
	Node
	expressionNode()
}

type Program struct {
	Statements []Statement
}

func (p *Program) TokenLiteral() string {
	if len(p.Statements) > 0 {
		return p.Statements[0].TokenLiteral()
	}
	return ""
}

type BlockStatement struct {
	Statements []Statement
}

func (bs *BlockStatement) statementNode()       {}
func (bs *BlockStatement) TokenLiteral() string { return "" }

type SetCommand struct {
	Variable string
	Value    Expression
}

func (s *SetCommand) statementNode() {}
func (s *SetCommand) TokenLiteral() string { return "set" }

type SetNicknameCommand struct {
	User     Expression
	Nickname Expression
}

func (s *SetNicknameCommand) statementNode() {}
func (s *SetNicknameCommand) TokenLiteral() string { return "set nickname" }

type IfStatement struct {
	Condition Expression
	Then      *BlockStatement
	Else      *BlockStatement
}

func (i *IfStatement) statementNode() {}
func (i *IfStatement) TokenLiteral() string { return "if" }


type StartBotCommand struct {
	Token Expression
}

func (s *StartBotCommand) statementNode() {}
func (s *StartBotCommand) TokenLiteral() string { return "start" }

type WhenBotStartsCommand struct {
	Body *BlockStatement
}

func (s *WhenBotStartsCommand) statementNode()       {}
func (s *WhenBotStartsCommand) TokenLiteral() string { return "when bot starts" }

type WhenMessageCommand struct {
	Condition Expression
	Body      *BlockStatement
}

func (s *WhenMessageCommand) statementNode()       {}
func (s *WhenMessageCommand) TokenLiteral() string { return "when message" }

type WhenMessageSentCommand struct {
	Body *BlockStatement
}

func (s *WhenMessageSentCommand) statementNode()       {}
func (s *WhenMessageSentCommand) TokenLiteral() string { return "when message is sent" }

type WhenMemberJoinsCommand struct {
	Body *BlockStatement
}

func (s *WhenMemberJoinsCommand) statementNode()       {}
func (s *WhenMemberJoinsCommand) TokenLiteral() string { return "when member joins" }

type WhenMemberLeavesCommand struct {
	Body *BlockStatement
}

func (s *WhenMemberLeavesCommand) statementNode()       {}
func (s *WhenMemberLeavesCommand) TokenLiteral() string { return "when member leaves" }

type WhenDiscordEventCommand struct {
	EventName Expression
	Body      *BlockStatement
}

func (s *WhenDiscordEventCommand) statementNode()       {}
func (s *WhenDiscordEventCommand) TokenLiteral() string { return "when discord event" }

type ReplyWithCommand struct {
	Message Expression
}

func (s *ReplyWithCommand) statementNode()       {}
func (s *ReplyWithCommand) TokenLiteral() string { return "reply with" }

type ReplyWithEmbedCommand struct {
	Title       Expression
	Description Expression
}

func (s *ReplyWithEmbedCommand) statementNode()       {}
func (s *ReplyWithEmbedCommand) TokenLiteral() string { return "reply with embed" }

type BanCommand struct {
	User   Expression
	Reason Expression
}

func (s *BanCommand) statementNode()       {}
func (s *BanCommand) TokenLiteral() string { return "ban user" }

type UnbanCommand struct {
	User   Expression
}

func (s *UnbanCommand) statementNode()       {}
func (s *UnbanCommand) TokenLiteral() string { return "unban user" }

type KickCommand struct {
	User   Expression
	Reason Expression
}

func (s *KickCommand) statementNode()       {}
func (s *KickCommand) TokenLiteral() string { return "kick user" }

type TimeoutCommand struct {
	User     Expression
	Duration Expression
}

func (s *TimeoutCommand) statementNode()       {}
func (s *TimeoutCommand) TokenLiteral() string { return "timeout user" }

type AddRoleCommand struct {
	Role Expression
	User Expression
}

func (s *AddRoleCommand) statementNode()       {}
func (s *AddRoleCommand) TokenLiteral() string { return "add role" }

type RemoveRoleCommand struct {
	Role Expression
	User Expression
}

func (s *RemoveRoleCommand) statementNode()       {}
func (s *RemoveRoleCommand) TokenLiteral() string { return "remove role" }

type CreateThreadCommand struct {
	Name Expression
}

func (s *CreateThreadCommand) statementNode()       {}
func (s *CreateThreadCommand) TokenLiteral() string { return "create thread" }

type PurgeCommand struct {
	Amount Expression
}

func (s *PurgeCommand) statementNode()       {}
func (s *PurgeCommand) TokenLiteral() string { return "purge messages" }

type StringLiteral struct {
	Value string
}

func (s *StringLiteral) expressionNode()      {}
func (s *StringLiteral) TokenLiteral() string { return s.Value }

type NumberLiteral struct {
	Value float64
}

func (n *NumberLiteral) expressionNode()      {}
func (n *NumberLiteral) TokenLiteral() string { return fmt.Sprintf("%v", n.Value) }

type Identifier struct {
	Value string
}

func (i *Identifier) expressionNode()      {}
func (i *Identifier) TokenLiteral() string { return i.Value }

type RegisterSlashCommand struct {
	Name        Expression
	Description Expression
	Options     Expression // Should resolve to an array of objects
}

func (s *RegisterSlashCommand) statementNode()       {}
func (s *RegisterSlashCommand) TokenLiteral() string { return "register slash command" }

type WhenCommandUsedCommand struct {
	CommandName Expression
	Body        *BlockStatement
}

func (s *WhenCommandUsedCommand) statementNode()       {}
func (s *WhenCommandUsedCommand) TokenLiteral() string { return "when command used" }

type MemberExpression struct {
	Object   Expression
	Property Expression
}

func (m *MemberExpression) expressionNode()      {}
func (m *MemberExpression) TokenLiteral() string { return "." }

type BinaryExpression struct {
	Left     Expression
	Operator string
	Right    Expression
}

func (b *BinaryExpression) expressionNode()      {}
func (b *BinaryExpression) TokenLiteral() string { return b.Operator }

type DeferReplyCommand struct{}

func (s *DeferReplyCommand) statementNode()       {}
func (s *DeferReplyCommand) TokenLiteral() string { return "defer reply" }

type EditReplyCommand struct {
	Message Expression
}

func (s *EditReplyCommand) statementNode()       {}
func (s *EditReplyCommand) TokenLiteral() string { return "edit reply with" }

type ReplyWithModalCommand struct {
	ModalID Expression
	Title   Expression
}

func (s *ReplyWithModalCommand) statementNode()       {}
func (s *ReplyWithModalCommand) TokenLiteral() string { return "reply with modal" }

type WhenModalSubmittedCommand struct {
	ModalID Expression
	Body    *BlockStatement
}

func (s *WhenModalSubmittedCommand) statementNode()       {}
func (s *WhenModalSubmittedCommand) TokenLiteral() string { return "when modal submitted" }


type DeleteMessageCommand struct{}

func (s *DeleteMessageCommand) statementNode()       {}
func (s *DeleteMessageCommand) TokenLiteral() string { return "delete message" }

type EditMessageCommand struct {
	NewText Expression
}

func (s *EditMessageCommand) statementNode()       {}
func (s *EditMessageCommand) TokenLiteral() string { return "edit message to" }

type CreateChannelCommand struct {
	Name Expression
	Type Expression
}

func (s *CreateChannelCommand) statementNode()       {}
func (s *CreateChannelCommand) TokenLiteral() string { return "create channel" }

type CreateRoleCommand struct {
	Name  Expression
	Color Expression
}

func (s *CreateRoleCommand) statementNode()       {}
func (s *CreateRoleCommand) TokenLiteral() string { return "create role" }

type DeleteRoleCommand struct {
	RoleID Expression
}

func (s *DeleteRoleCommand) statementNode()       {}
func (s *DeleteRoleCommand) TokenLiteral() string { return "delete role" }

type DeleteChannelCommand struct {
	ChannelID Expression
}

func (s *DeleteChannelCommand) statementNode()       {}
func (s *DeleteChannelCommand) TokenLiteral() string { return "delete channel" }

type JoinVoiceCommand struct {
	ChannelID Expression
}

func (s *JoinVoiceCommand) statementNode()       {}
func (s *JoinVoiceCommand) TokenLiteral() string { return "join voice channel" }

type PlayAudioCommand struct {
	AudioURL Expression
}

func (s *PlayAudioCommand) statementNode()       {}
func (s *PlayAudioCommand) TokenLiteral() string { return "play audio" }

type ReplyWithAttachmentCommand struct {
	FilePath Expression
}

func (s *ReplyWithAttachmentCommand) statementNode()       {}
func (s *ReplyWithAttachmentCommand) TokenLiteral() string { return "reply with file" }


type IfUserHasPermissionStatement struct {
        Permission Expression
        Body       *BlockStatement
}

func (s *IfUserHasPermissionStatement) statementNode()       {}
func (s *IfUserHasPermissionStatement) TokenLiteral() string { return "if user has permission" }

type CreateAutoModRuleCommand struct {
        Name          Expression
        KeywordFilter Expression
}

func (s *CreateAutoModRuleCommand) statementNode()       {}
func (s *CreateAutoModRuleCommand) TokenLiteral() string { return "create automod rule" }

type SendPollCommand struct {
	Title   Expression
	Options []Expression
}
func (s *SendPollCommand) statementNode()       {}
func (s *SendPollCommand) TokenLiteral() string { return "reply with poll" }

type CreateScheduledEventCommand struct {
	Name      Expression
	StartTime Expression
}
func (s *CreateScheduledEventCommand) statementNode()       {}
func (s *CreateScheduledEventCommand) TokenLiteral() string { return "create scheduled event" }

type CreateEmojiCommand struct {
	Name Expression
	Path Expression
}
func (s *CreateEmojiCommand) statementNode()       {}
func (s *CreateEmojiCommand) TokenLiteral() string { return "create emoji" }

type ArrayLiteral struct {
        Elements []Expression
}

func (a *ArrayLiteral) expressionNode()      {}
func (a *ArrayLiteral) TokenLiteral() string { return "[" }

type ObjectLiteral struct {
	Pairs map[string]Expression
}

func (o *ObjectLiteral) expressionNode()      {}
func (o *ObjectLiteral) TokenLiteral() string { return "{" }


type CreateWebhookCommand struct {
	Name Expression
}
func (s *CreateWebhookCommand) statementNode()       {}
func (s *CreateWebhookCommand) TokenLiteral() string { return "create webhook" }
type SendViaWebhookCommand struct {
	WebhookID    Expression
	WebhookToken Expression
	Content      Expression
}

func (s *SendViaWebhookCommand) statementNode()       {}
func (s *SendViaWebhookCommand) TokenLiteral() string { return "send via webhook" }

type SendDMCommand struct {
	TargetID Expression
	Message  Expression
}

func (s *SendDMCommand) statementNode()       {}
func (s *SendDMCommand) TokenLiteral() string { return "send dm" }


type ReplyWithStickerCommand struct {
	StickerID Expression
}
func (s *ReplyWithStickerCommand) statementNode()       {}
func (s *ReplyWithStickerCommand) TokenLiteral() string { return "reply with sticker" }

type GetBotInfoCommand struct {}
func (s *GetBotInfoCommand) statementNode()       {}
func (s *GetBotInfoCommand) TokenLiteral() string { return "get bot info" }
